keps/sig-storage/2857-runtime-assisted-pv-mounts/README.md

```diff
@@ -0,0 +1,2051 @@
+reporting errors when checking for the `DEFER_FS_OPS` capability of the CSI
+plugin prior to invoking CSI APIs: `NodeGetVolumeStats` and `NodeExpandVolume`.
+
+When the pod terminates, the container runtime dismounts the file system on the
+block devices in the micro VM in response to CRI `StopPodSandbox`. Kubelet will
+invoke `NodeUnpublishVolume` without specifying `defer_fs_ops` (due to lack of
+`DEFER_FS_OPS` capability of the downgraded CSI plugin). The downgraded CSI
+plugin will find that no file system mount is present on the host and thus there
+is nothing to clean up. The CSI plugin will return success to Kubelet in
+accordance with the idempotent requirements of repeated CSI calls. The file
+system metadata file that was generated earlier will need to be cleaned up by
+the Kata runtime.
+
+##### Story 6: Fail pod startup due to CSI plugin version downgrade right after Kubelet began pod initialization
+
+All assumptions for the cluster mentioned earlier are true except the CSI plugin
+`plugin1.example.com` gets downgraded from a version that did support deferral
+of file system operations to a version that does not support deferral. A pod
+(with `runtimeClassName` set to `microvm` and PVCs with StorageClass set to
+`fast-storage`) got successfully scheduled right before config changes to
+reflect the plugin downgrade was initiated. Therefore, the scheduler plugin sets
+`DeferFSMount` in the `PersistentVolumeClaimVolumeSource`s of the pod spec to
+"true" (as the scheduler configuration still indicated `fast-storage`
+StorageClass capable of deferral of file system operations).
+
+When Kubelet begins to initialize volumes for the pod, it will notice that the
+CSI  plugin does not report `DEFER_FS_OPS` capability but the `DeferFSMount`
+field is set. This will result in failure to initialize the pod and reporting of
+FailedMount events on the pod. Kubelet will continue to try to perform the mount
+with backoff until the CSI plugin is upgraded to a version where it reports
+`DEFER_FS_OPS` capability or the pod is deleted.
+
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+There are certain design/implementation constraints derived from feedback in
+earlier iterations of this KEP that has shaped the current proposal:
+- Fields in the RuntimeClass or capabilities associated with a container runtime
+should not influence the behavior of the Kubelet around how it interacts with
+CRI (or OCI) plugins to bring up a pod. In the context of this KEP, it implies
+that the Kubelet should not query the capabilities of a container runtime
+handler to determine whether it is capable of deferred file system operations
+and perform any action based on such capabilities.
```

Comment:
I'm interpreting this as it is allowed for fields in RuntimeClass to affect how volumes are mounted / how NodeStage + NodePublish is called, as long as the resulting CRI call to start a pod / container is still the same. If that's true, we could add a field to RuntimeClass and *not spill* container runtime capabilities among Pods.

Reply:
Yes, having a set of fields in RuntimeClass declare the capabilities of a container runtime handler with regards to handling file system mount/management, is one alternative. However, this would involve spilling (sig-)storage concepts, especially around the ability of a container runtime handler's ability to apply specific post-mount configurations, to (sig-)node owned areas like RuntimeClass. A specific example is if new fsGroupChangePolicy fields are introduced in the future, that will require sig-node approval to introduce it as a potentially supported setting in the RuntimeClass fields (besides as a CSI plugin capability in the context of runtime assisted mounting).

To avoid having to involve sig-node for any new mount/post-mount capabilities, I am proposing we keep this as an opt-in field in the pod spec.

Reply:
To me it seems we're trying to work around communication between teams with worse design. With a field in Pod, cluster needs to deploy a Pod webhook that needs to _"[match] the runtimeClass of the pod with a set of configured runtimeClasses (that are capable of coordinating mount and management operations of file systems in PVCs mounted by the pod)"_.

So instead of a field in RuntimeClass, there needs to be a webhook that _knows_ what RuntimeClasses support runtime assisted mount. That knowledge will be most likely an annotation or label on the RuntimeClass anyway.

Reply:
Agree 100% - I really like the idea of the RuntimeClass being able to express its capabilities directly. Will explore this with sig-node.

Reply:
+1 on @jsafrane above.

Actually that is how GKE enable gVisor runtime initially. @Random-Liu to comment on this.

Reply:
Sounds good. I will restructure the KEP to remove the pod field and drive it using runtimeClass fields and have Kubelet automatically match runtimeClass with a storageClass. One factor to consider is: having Kubelet make the decision to automatically perform runtime-assisted mounts will require matching a set of file systems specified as supported in the runtimeClass with a set of file systems that the CSI plugin reports as supported. I will consider having the CSI plugins report the supported set of file systems through a new field in CSI `NodeGetInfoResponse` (as CSI `NodeGetCapabilitiesResponse` only supports boolean flags).

Reply:
Actually in the initial `RuntimeClass` design, it does include the capabilities in the scope. https://docs.google.com/document/d/1d5OG5z0fb6JZ5BHDToam9170RXx4FL1q8gYpX7r_UAU/edit#

There are other capabilities it can express, e.g. whether `privileged` pod is supported, whether `apparmor` is supported etc.

Reply:
Thanks Lantao for providing the details. Yes, that is what we originally designed for, but later narrowed down the scope. I shared this at SIG Node meeting this Tuesday, but failed to find the original doc. cc/ @tallclair the original author of RuntimeClass.